### PR TITLE
Don't allow empty username for git

### LINKED ISSUE #1926: Unable to connect & Download from Cloud Storage on RaspberryPi
### Describe the issue:

- [X] Bug
- [ ] Change request
- [ ] New feature request
- [ ] Discussion request

### Issue long description:
Compilation of software completed on a RaspberryPi (as no existing distribution available). Software starts up ok. I can download software from my dive computer but am unable to connect to cloud storage  due errors raised (see below). 

### Operating system:
Debian Stretch
RaspberryPi
Have tried twice (including on a clean installation). Same error is reported. 
I also run 4.8.4 on iOS with no problems (but this is not a self compiled version). 

### Subsurface version:
Version Built: 4.8.4
Does not happen on pre-built versions of 4.8.4 (I run the packages on iOS & OSX without problem). 
Self Compile
Git Commit Ref used: 0903bef9db8659313c4bb533c295505c0c3851bf

### Steps to reproduce:
Execute the software. 
Configure the Cloud Storage account in Preferences
Select File>Open Cloud Storage

### Current behavior:
On first execution the error message ‘Creating Commit Failed’
On second execution the error message ‘unable to lookup revision <email>’ is displayed

### Expected behavior:
Dive Log already stored in the cloud is downloaded and displayed.

### Additional information:
Build Log available if needed.

### PR DESCRIPTION
**Pull request by atdotde**
When no real name is set in /etc/passwd the username ends
up being the empty string. Git does not like that.

Before we only testing for the username being a NULL pointer.

Fixes #1926 

Reported-by: Keith Grimes
Signed-off-by: Robert C. Helling <helling@atdotde.de>

### Describe the pull request:

- [x] Bug fix
- [ ] Functional change
- [ ] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### CODE REVIEW

**Comment by atdotde** [comment]
@keithgrimes could you please check that this solves the issue?

**Review by keithgrimes — COMMENTED** [review]
Alas this does not fix the issue. The extended attributes in the /etc/passwd file (as changed by chfn) are multiple and from what I can see are comma delimited (permitting 4 entries). As such the field returned is not blank but instead ',,,' (as seen in the log file). If it helps the entry in /etc/passwd is 'pi:x:1000:1000:,,,:/home/pi:/bin/bash'. 

The log file highlights this with the line (values of the fields in brackets)
git storage, enter get_signature_now user.name(,,,) user.email(pi@pimobile)
user.email is build separately from username and hostname, not from the passwd struct.

I assume the call to pwd->pw_gecos in unix.c/subsurface_user_info is not able to distinguish when a full name has not been set. I assume there are good reasons why the username field is not used?

However, not sure if checking against ',,,' would be a good/safe check as it depends if other fields are set, but not the full name. This is where my linux experience is not sufficient to understand the potential impact. 

Regards
Keith

**Comment by keithgrimes** [comment]
Hi atdotde, 

Glad you have reproduced. Given the field covers multiple items (including office, work phone, home phone) I did not know if the other fields were populated (but not full name) what would happen and this would not match ‘,,,’. Unlikely I would have thought, but one to think about. 

As for my goal, I think you are spot on. I run subsurface on a Macbook Pro, iPad mini 2, iPhone and now my new iPad Pro. However, when travelling my Macbook is quite heavy (old 2011 aluminium body one) and hence causes some issues with luggage but this is the only one with USB connectivity for my Suunto Zoop Novo for downloading dives while I am away. 

With the Pi now up and running I can download dives to the Pi from my DC, upload them to the Cloud and either edit on the Pi (using a terminal session) or use my iPad Pro.

Please note for this I do run 2 network adapters (one as a wifi hotspot, the other as a link to the internet) with ip fowarding enabled. 

As for your plan, happy to help while when I have time. Please be aware though (as you will have seen from my questions) I have not ‘formally’ developed for a number of years (20+) and even then I was Microsoft based. I have played with Linux but I would still class myself as on the basics. But with 30+ years in the development industry (and Google) nothing takes too long to master.

Regards

Keith.

**Comment by keithgrimes** [comment]
If you have some details on your project (GitHub reference?) let me know and I will have a look.

**Review by dirkhh — CHANGES_REQUESTED** [review]
It passes the build and test, but it doesn't do what you want it to do, I think.

File: core/save-git.c [hunk 1]
```diff
@@ -1034,7 +1034,7 @@ static int get_authorship(git_repository *repo, git_signature **authorp)
 	/* try to fetch the user info from the OS, otherwise use default values. */
 	struct user_info user = { .name = NULL, .email = NULL };
 	subsurface_user_info(&user);
-	if (!user.name)
+	if (!user.name || empty_string(user.name || same_string(user.name, ",,,")))
```
[the remark this answers is not in the document]
Comment (atdotde) [reply core/save-git.c]: You just won a bet.

**Comment by atdotde** [comment]
I think rather that dealing with special strings like ",,,", we should rather just take the actual name and discard anything after the first comma (this should be info about office and phone number I guess). See updated version.

**Comment by atdotde** [comment]
@keithgrimes I made this a separate issue #1935.

File: core/save-git.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (bstoeger) [reply core/save-git.c]: Also note that `empty_string(s)` already checks for `!s`. So the whole `!user.name` expression is not necessary, though it doesn't hurt either.

File: core/unix.c [hunk 2]
```diff
@@ -38,8 +38,16 @@ void subsurface_user_info(struct user_info *user)
...
 
 	if (pwd) {
-		if (!empty_string(pwd->pw_gecos))
+		if (!empty_string(pwd->pw_gecos)) {
 			user->name = strdup(pwd->pw_gecos);
+			// We only want the name, not the office or phone number
+			char *c = user->name;
+			while (*c) {
+				if (*c == ',')
+					*c = '\0';
```
Comment (bstoeger) [inline core/unix.c:47]: `break` the loop here?

**Comment by keithgrimes** [comment]
Checked this on my build (without the break additional break) and save works successfully.

File: core/unix.c (hunk 2 above)
Comment (atdotde) [reply core/unix.c:47]: Updated. Also same_string against "" is overkill.

File: core/save-git.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (atdotde) [reply core/save-git.c]: I have replaced the same_string as when testing against "" it is overkill.

**Comment by bstoeger** [comment]
Looks correct (can a Unix username contain a ','?) - commit.

**Pull request merged**